keps/sig-api-machinery/4020-unknown-version-interoperability-proxy/README.md

```diff
@@ -0,0 +1,941 @@
+
+We should ensure at most one proxy, rather than proxying the request over and over again (if the source apiserver has an incorrect understanding of what the destination apiserver can serve).
+
+To prevent server-side request forgeries we will not give control over information about apiserver IP/endpoint and the trust bundle (used to authenticate server while proxying) to users via REST APIs.
+
+## Design Details
+
+### Aggregation Layer
+
+1. A new filter will be added to the [handler chain] of the aggregation layer. This filter will maintain an internal map with the key being the group-version-resource and the value being a list of server IDs of apiservers that are capable of serving that group-version-resource
+   1. This internal map is populated using an informer for StorageVersion objects. An event handler will be added for this informer that will get the apiserver ID of the requested group-version-resource and update the internal map accordingly
+
+2. This filter will pass on the request to the next handler in the local aggregator chain, if:
+   1. It is a non resource request
+   2. The StorageVersion informer cache hasn't synced yet or if `StorageVersionManager.Completed()` has returned false. We will serve error 503 in this case
+   3. The request has a header that indicates that this request has been proxied once already. If for some reason the resource is not found locally, we will serve error 503
+   4. No StorageVersion was retrieved for it, meaning the request is for an aggregated API or for a custom resource
+   5. If the local apiserver ID is found in the list of serviceable-by server IDs from the internal map
+
+3. If the local apiserver ID is not found in the list of serviceable-by server IDs, a random apiserver ID will be selected from the retrieved list and the request will be proxied to this apiserver
+
+4. If there is no apiserver ID retrieved for the requested GVR, we will serve 404 with error `GVR <group_version_resource> is not served by anything in this cluster`
+
+5. If the proxy call fails for network issues or any reason, we serve 503 with error `Error while proxying request to destination apiserver`
+
+[handler chain]:https://github.com/kubernetes/kubernetes/blob/fc8f5a64106c30c50ee2bbcd1d35e6cd05f63b00/staging/src/k8s.io/apiserver/pkg/server/config.go#L639
+
+#### StorageVersion enhancement needed
+
+StorageVersion API currently tells us whether a particular StorageVersion can be read from etcd by the listed apiserver. We will enhance this API to also include apiserver ID of the server that can serve this StorageVersion.
+
+#### Identifying destination apiserver's network location
+
+We will use the already existing [masterlease reconciler](https://github.com/kubernetes/kubernetes/blob/master/pkg/controlplane/reconcilers/lease.go) to store/retrieve the IPs and ports for kube-apiservers. Major reasons to use this are
```

Comment:
I still have doubts about reusing the existing reconciler, this is the default reconciler but is optional, it depends on the `EndpointReconcilerType` chosen. I was thinking on having a totally new controller despite most of the code will be duplicated, but there will be zero risk of affecting this core logic.

The other option I can see is to use the existing v1.Lease controller instead of the etcd Leases

```
	m.GenericAPIServer.AddPostStartHookOrDie("start-kube-apiserver-identity-lease-controller", func(hookContext genericapiserver.PostStartHookContext) error {
			kubeClient, err := kubernetes.NewForConfig(hookContext.LoopbackClientConfig)
			if err != nil {
				return err
			}

			// generate a context  from stopCh. This is to avoid modifying files which are relying on apiserver
			// TODO: See if we can pass ctx to the current method
			ctx := wait.ContextForChannel(hookContext.StopCh)

			leaseName := m.GenericAPIServer.APIServerID
			holderIdentity := m.GenericAPIServer.APIServerID + "_" + string(uuid.NewUUID())

			controller := lease.NewController(
				clock.RealClock{},
				kubeClient,
				holderIdentity,
				int32(IdentityLeaseDurationSeconds),
				nil,
				IdentityLeaseRenewIntervalPeriod,
				leaseName,
				metav1.NamespaceSystem,
				// TODO: receive identity label value as a parameter when post start hook is moved to generic apiserver.
				labelAPIServerHeartbeatFunc(KubeAPIServer))
			go controller.Run(ctx)
```
/cc @wojtek-t

Reply:
@deads2k how do you feel about using the APIServerIdentity v1.Lease to store/retrieve the apiserver IP info?

Reply:
> @deads2k how do you feel about using the APIServerIdentity v1.Lease to store/retrieve the apiserver IP info?

I think this would work well and I think the security profile on it is ok since cluster-admins will  be able to specify an egress dialer.  You should confirm with @jpbetz

Reply:
>Does this egress dialer support make this safe to run in a cloud provider? i.e. we cannot have cluster admins using this to make network request against backend networks accessible to the Kube API server.

I haven't ever wired it personally, but my understanding is that it can be used to restrict network access. @jkh52 would be on my list to ask.